claat/serve.go

```diff
@@ -72,214 +60,232 @@ func cmdServe() {
-	"stacky":                        "PolymerElements/stacky#1.3.2",
-	"webcomponentsjs":               "WebComponents/webcomponentsjs#0.7.24",
+	return ioutil.WriteFile(name, content, 0644)
 }
 
-// overrideComp maps a given spec, as described in fetchRepo doc comments to a
-// desired component name.
-var overrideComp = map[string]string{
-	"googlecodelabs/codelab-components": "google-codelab-elements",
-	"google/code-prettify":              "google-prettify",
-}
+// warnOnce makes sure we notify users about fetching dependencies only once.
+var warnOnce sync.Once
 
-// fetchRepo downloads a repo from github and unpacks it into basedir/dest,
+// fetchRepo downloads a repo from github and unpacks it into basedir/name,
 // calling itself recursively for every dependency found in the unpacked bower.json.
-// If the basedir/dest directory already exists, fetchRepo does nothing.
+// If the basedir/name directory already exists, fetchRepo does nothing.
 //
 // The spec has the following format: "githubuser/repo#version".
 // Any non-alphanumeric characters are stripped from the version.
-// If version is missing, "master" is used instead.
-// The bower registry is not used for naming resolution.
-func fetchRepo(basedir, spec string) error {
-	var user, repo, path, comp, ver string
-	s := strings.Split(spec, "#")
-	path = s[0]
-	if len(s) > 1 {
-		ver = s[1]
-	}
-	ver = strings.Trim(ver, "^~ ")
-	if ver == "" {
-		ver = "master"
-	}
-	s = strings.Split(path, "/")
-	user = s[0]
-	if len(s) > 1 {
-		repo = s[1]
-	}
-	// Check exception map to see if we should override the component name.
-	comp = repo
-	if c, ok := overrideComp[path]; ok {
-		comp = c
-	}
-	// if repo already exists locally, return immediately, we're done.
-	if _, err := os.Stat(basedir + "/" + comp); err == nil {
+// If version is missing, "master" is used instead, unless overridden in bowerVersionOverride.
+// Instead of bower registry a local bowerSpecResolve map is used for naming resolution.
+func fetchRepo(basedir, name, spec string) error {
+	outdir := filepath.Join(basedir, name)
+	if _, err := os.Stat(outdir); err == nil {
 		return nil
 	}
-	zipFile := basedir + "/" + comp + ".zip"
-	url := "https://github.com/" + user + "/" + repo + "/archive/v" + ver + ".zip"
-	err := downloadFile(zipFile, url)
+
+	warnOnce.Do(func() {
+		log.Println("Fetching dependencies...")
+	})
+	tryURL, err := fromBowerSpec(name, spec)
 	if err != nil {
-		os.Remove(zipFile)
-		url = "https://github.com/" + user + "/" + repo + "/archive/" + ver + ".zip"
-		err = downloadFile(zipFile, url)
-		if err != nil {
-			return err
+		return err
+	}
+	zipFile := filepath.Join(basedir, name+".zip")
+	var ok bool
+	for _, u := range tryURL {
+		if err := downloadFile(zipFile, u); err == nil {
+			ok = true
+			break
 		}
 	}
-	err = unzip(zipFile, basedir)
-	if err != nil {
+	if !ok {
+		return fmt.Errorf("fetchRepo(%q, %q): could not find in any of %q", name, spec, tryURL)
```

Comment:
I know %q escapes strings but will it do the right thing in terms of formatting tryURL, which is array of strings? Or would you want to use %v for that arg?